### DISCUSSION TITLE
celery.loaders.base.autodiscover_tasks should re-raise non-task ImportErrors

### DISCUSSION THREAD

**Discussion by johnjameswhitman** [celery/celery#8620] [General]
# Problem

We ran into an issue yesterday where Celery-on-Django was not able to auto-discover a task in our deployed environments, but was able to discover it locally. It turned out that locally the team all had a native system package installed that supplied the `libmagic` shared library for `python-magic`, but that wasn't present in our docker images. I was finally able to figure out what was going on by throwing some exception logging into `celery.loaders.base.find_related_module` here and observing what package the `ImportError` was complaining about:

https://github.com/celery/celery/blob/ae54d4100f976bf4ed4d12544d8acf80803d9cf3/celery/loaders/base.py#L266-L272

It seems like this is meant to swallow errors related to importing the task-module-at-hand, and then re-raise `ImportErrors` for other packages / modules. That seems fine, but it didn't work in our case -- I believe the `import_exc_name` was the task's package-name so it just returned `None` on L272.

The result was befuddling to us, because there was no clear indication of _why_ the tasks weren't being auto-discovered (and indeed they were after we baked the native package / dependency into our Dockerfile). It would've been helpful if Celery just crashed when the `libmagic` import wasn't found.

# Idea

When debugging I noticed that `importlib` actually raises a `ModuleNotFoundError` when it can't autodiscover tasks in a package, which is a subclass of `ImportError` (example below). Would it make sense to only swallow that narrower error, and let others bubble up? I'm not sure if that code is indeed meant to catch the broader `ImportError`.

```
Traceback (most recent call last):
  File "/app/config/celery_loader.py", line 290, in find_related_module
    return importlib.import_module(module_name)
  File "/usr/local/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 984, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'api.some_app.tasks'
```

**Comment by auvipy** [comment]
thanks for the report and figuring out the root cause. It would be great if you could come with a draft solution and we can continue discussion there?

**Reply by johnjameswhitman** [reply]
Adding a few more notes below. I should be able to draft up a PR tomorrow.

After playing with this some more, this appears to happen when the error originates from outside of the `importlib.import_module` call-stack described below.
- The root cause appears to be two-fold:
	- [`magic.loader.load_lib`](https://github.com/ahupp/python-magic/blob/54d86fd61b5d655e6976115a786c77d646e50341/magic/loader.py#L49) raises an `ImportError` but does not set a `name` attribute like `importlib.import_module` does (see call-stack for latter below).
		- [`importlib.import_module`](https://github.com/python/cpython/blob/289af8612283508b67d7969d7182070381b4349b/Lib/importlib/__init__.py#L88)
		- [`importlib._bootstrap._gcd_import`](https://github.com/python/cpython/blob/289af8612283508b67d7969d7182070381b4349b/Lib/importlib/_bootstrap.py#L1381)
		- [`importlib._bootstrap._find_and_load`](https://github.com/python/cpython/blob/289af8612283508b67d7969d7182070381b4349b/Lib/importlib/_bootstrap.py#L1364)
			- this raises and sets the `name` itself
			- it also calls out to a few places that do the same
	- [`celery.loaders.base.find_related_module` defaults the error's `name`](https://github.com/celery/celery/blob/ae54d4100f976bf4ed4d12544d8acf80803d9cf3/celery/loaders/base.py#L269) attribute to the task-module being searched.
	- So, when `magic.loader.load_lib` raises its `ImportError` without a `name`, it gets ignored.

Celery's behavior as introduced in https://github.com/celery/celery/pull/5211. Outside of this narrow circumstance, I think the code works-as-intended (it will catch non-existent imports that `importlib` tries to follow).

**Reply by auvipy** [reply]
thanks a lot for digging deep into it. that was a very old PR so I really can't recall much. so a new PR with better approach will be really helpful. also the eco system has changed and our python knowledge improves.

**Reply by johnjameswhitman** [reply]
Dropped a PR w/ some tests here: https://github.com/celery/celery/pull/8632
